### PR TITLE
Fix prefix probe so it doesn't use a cursor

### PR DESCRIPTION
**Pull request by gsramdasi**
Signed-off-by: Gaurav Ramdasi <10132364+gsramdasi@users.noreply.github.com>

## Description
Prefix probes do not use cursors underneath.
The prefix can be any length.
KVS cparam for suffix length is removed.
Prefix probe is now a stable API.

## Issue(s) Addressed
NFSE-5171: remove suffix tree configuration parameter
NFSE-5212: Change prefix probe to use a probe instead of a cursor
~NFSE-5250: Stabilize prefix probe API~

### CODE REVIEW

File: lib/binding/kvdb_interface.c [hunk 1]
```diff
@@ -1254,33 +1254,40 @@ hse_kvs_prefix_probe(
...
 
-    *found = HSE_KVS_PFX_FOUND_MUL;
     PERFC_INCADD_RU(&kvdb_pc, PERFC_RA_KVDBOP_KVS_PFXPROBE, PERFC_RA_KVDBOP_KVS_GETB,
-                    klen + *key_len + *val_len);
+                    *key_len + *val_len);
 
-out:
-    ikvdb_kvs_cursor_destroy(cur);
-    return err;
+    return 0;
```
Comment (beckerg) [inline lib/binding/kvdb_interface.c:1290]: Can you trust that key_len is not NULL?

File: lib/cn/cn_tree.c [hunk 2]
```diff
@@ -820,6 +820,7 @@ cn_tree_lookup(
     err = 0;
 
     while (node) {
+        struct route_node *rn = node->tn_route_node;
```
Comment (beckerg) [inline lib/cn/cn_tree.c:823]: Move this down to where you need use it..

File: lib/cn/cn_tree.c [hunk 3]
```diff
@@ -841,16 +842,34 @@ cn_tree_lookup(
...
+
+            node = cn_tree_node_lookup(tree, kt->kt_data, kt->kt_len);
+        }
     }
 
   done:
     rmlock_runlock(lock);
 
-    if (qctx) {
+    if (is_probe) {
```
Comment (beckerg) [inline lib/cn/cn_tree.c:872]: I am think we should split this function into two, one for normal, and one for probe...

File: lib/cn/route.h [hunk 4]
```diff
@@ -141,11 +141,18 @@ route_node_key_modify(
...
 {
-    return keycmp(node->rtn_keybufp, node->rtn_keylen, key, klen);
+    return keycmp(key, klen, node->rtn_keybufp, node->rtn_keylen);
 }
 
+static HSE_ALWAYS_INLINE int
+route_node_keycmp_prefix(const void *pfx, uint pfxlen, const struct route_node *node)
+{
+    return keycmp_prefix(pfx, pfxlen, node->rtn_keybufp, node->rtn_keylen);
+}
```
Comment (beckerg) [inline lib/cn/route.h:153]: route_node * is the first argument to most of the route_node_*() functions, why did you move it?

**Review by beckerg — APPROVED** [review]

File: include/hse/hse.h [hunk 5]
```diff
@@ -596,6 +596,62 @@ hse_kvs_get(
...
+    const void *                pfx,
+    size_t                      pfx_len,
+    enum hse_kvs_pfx_probe_cnt *found,
+    void *                      keybuf,
+    size_t                      keybuf_sz,
+    size_t *                    key_len,
+    void *                      valbuf,
+    size_t                      valbuf_sz,
+    size_t *                    val_len);
+
```
Comment (alexttx) [inline include/hse/hse.h:654]: Keep this experimental for now.  See comments in NFSE-5250.

File: lib/cn/cn_tree.c (hunk 3 above)
Comment (gsramdasi) [reply lib/cn/cn_tree.c:872]: Yes, makes sense.

File: lib/cn/route.h (hunk 4 above)
Comment (gsramdasi) [reply lib/cn/route.h:153]: Yeah we do, but I felt it would be more natural to compare the key under consideration against the edge key. Especially in the case of the prefix key compare where the prefix is compared against the edge key's prefix.

File: lib/kvs/kvs.c [hunk 6]
```diff
@@ -531,23 +478,12 @@ kvs_pfx_probe(
     struct lc *       lc = kvs->ikv_lc;
     struct cn *       cn = kvs->ikv_cn;
     uintptr_t         seqnoref = 0;
-    struct query_ctx  qctx;
+    struct query_ctx  qctx = {0};
```
Comment (tristan957) [inline lib/kvs/kvs.c:481]: spaces around the zero!

**Review by tristan957 — COMMENTED** [review]
Can you make sure hse-python and hse-java don't need updates?

File: lib/binding/kvdb_interface.c (hunk 1 above)
Comment (gsramdasi) [reply lib/binding/kvdb_interface.c:1290]: Nope. Good point. Fixed.

File: lib/cn/route.h (hunk 4 above)
Comment (beckerg) [reply lib/cn/route.h:153]: Not following, why not keep the route_node API consistent?

Comment (gsramdasi) [reply lib/cn/route.h:153]: Look at `keycmp_prefix(A, B)`. This compares A with B's prefix. Calling `route_node_keycmp_prefix(node, A)` may cause confusion if we're comparing A with the edge key's prefix or the edge key with A's prefix. So I thought it'd be consistent with `keycmp_prefix()` to have (A, node) as the arguments.

The `route_node_keycmp()` API could stay as it was, but I changed it so it's consistent with `route_node_keycmp_prefix()`.

Tbh, I don't feel too strongly about it. If it doesn't make sense to people, I can use the (node, A) order or arguments in both the calls.

Comment (beckerg) [reply lib/cn/route.h:153]: Gotcha..  Something weird about this API being part of route-map, can't put my finger on it...  :-)

**Comment by gsramdasi** [comment]
> 

They do. And in fact they need to go in first so the checks pass for this PR.
I have PRs up for them, but forgot to add reviewers!

https://github.com/hse-project/hse-python/pull/64
https://github.com/hse-project/hse-java/pull/38

**Comment by gsramdasi** [comment]
> Can you make sure hse-python and hse-java don't need updates?

They do. And in fact they need to go in first so the checks pass for this PR.
I have PRs up for them, but forgot to add reviewers!

https://github.com/hse-project/hse-python/pull/64
https://github.com/hse-project/hse-java/pull/38

File: lib/cn/route.h (hunk 4 above)
Comment (gsramdasi) [reply lib/cn/route.h:153]: One option could be to have a non-copying version of `route_node_keycpy()` to just get a pointer to the edge key and its length. Then the caller can do `keycmp()` or `keycmp_prefix()` and we'd avoid incurring any unnecessary key copy.

But the onus is on the caller to make sure the edge key pointer is not used outside the tree lock section.

File: lib/kvs/query_ctx.c [hunk 7]
```diff
@@ -3,219 +3,33 @@
  * Copyright (C) 2015-2020 Micron Technology, Inc.  All rights reserved.
```
Comment (tristan957) [inline lib/kvs/query_ctx.c:3]: Update copyright

File: lib/kvs/query_ctx.c [hunk 8]
```diff
@@ -3,219 +3,33 @@
...
-    u64              hash = hse_hash64(sfx, sfx_len);
-
-    assert(sfx_len > 0);
-    root = &qctx->tomb_tree;
-    link = &root->rb_node;
-    parent = NULL;
-
-    while (*link) {
-        struct tomb_elem *te;
+    uint64_t hash = hse_hash64(key, klen);
```
Comment (tristan957) [inline lib/kvs/query_ctx.c:15]: const

File: tests/unit/cn/cn_tree_test.c [hunk 9]
```diff
@@ -453,7 +453,7 @@ MTF_DEFINE_UTEST_PRE(test, t_simple_api, test_setup)
 
     struct cn_tree *    tree = 0;
     struct kvs_cparams *out,
-        cp = {.sfx_len = 0, .pfx_len = 12, };
+        cp = {.pfx_len = 12, };
```
Comment (tristan957) [inline tests/unit/cn/cn_tree_test.c:456]: space before the dot and lose the inline trailing comma

File: tests/unit/kvs/kvs_cparams_test.c [hunk 10]
```diff
@@ -78,25 +78,6 @@ MTF_DEFINE_UTEST_PRE(kvs_cparams_test, kvs_ext01, test_pre)
     ASSERT_EQ(UINT32_MAX, ps->ps_bounds.as_uscalar.ps_max);
 }
 
-MTF_DEFINE_UTEST_PRE(kvs_cparams_test, suffix_length, test_pre)
```
Comment (tristan957) [inline tests/unit/kvs/kvs_cparams_test.c:81]: Why remove the test if the rparam still exists?

File: tools/upsert/upsert.c [hunk 11]
```diff
@@ -0,0 +1,377 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright (C) 2022 Micron Technology, Inc.  All rights reserved.
+ */
+
+#include <endian.h>
```
Comment (tristan957) [inline tools/upsert/upsert.c:6]: Alphabetical please!

File: tools/upsert/upsert.c [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+
+        printf("%8u %12lu %8lu %8u\n", second, curr_opcnt, curr_opcnt - last_opcnt, test_ctx.curr_lap);
+
+        last_opcnt = curr_opcnt;
+        usleep(999 * 1000);
+    }
+}
+
+void
+syntax(const char *fmt, ...)
```
Comment (tristan957) [inline tools/upsert/upsert.c:220]: Use HSE_PRINTF

File: tools/upsert/upsert.c [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+    case EINVAL:
+        fatal(0, "missing group name (e.g. %s) before parameter %s\n",
+              PG_KVDB_OPEN, argv[optind]);
+        break;
+    default:
+        fatal(rc, "error processing parameter %s\n", argv[optind]);
+        break;
+    }
+
+    rc = rc ?: svec_append_pg(&hse_gparms, pg, PG_HSE_GLOBAL, NULL);
```
Comment (tristan957) [inline tools/upsert/upsert.c:345]: GNU-ism, ?:

File: tools/upsert/upsert.c (hunk 12 above)
Comment (gsramdasi) [reply tools/upsert/upsert.c:220]: Nice. I wasn't aware of this. Fixed.

File: tests/unit/kvs/kvs_cparams_test.c (hunk 10 above)
Comment (gsramdasi) [reply tests/unit/kvs/kvs_cparams_test.c:81]: I had removed this as a cparam. But then we decided to add it back as an rparam. So this test needs to go. But I've added back a test for the rparam.

File: lib/kvs/kvs_rparams.c [hunk 14]
```diff
@@ -217,6 +217,27 @@ static const struct param_spec pspecs[] = {
             },
         },
     },
+    {
+        .ps_name = "kvs_sfx_len",
```
[the remark this answers is not in the document]
Comment (tristan957) [reply lib/kvs/kvs_rparams.c:221]: Can you submit a new PR for reverting the name to what is used to be?
